src/py/mains/run.py

```diff
@@ -111,31 +120,141 @@ def wait_for_file(filename, task, job_name = ""):
             fc_run_logger.debug( " job: %r exited." % (job_name) )
             if not os.path.exists(filename):
                 fc_run_logger.warning( "%r is missing. job: %r failed!" % (filename, job_name) )
-            break
+            return Waiter.FoundExit
         if os.path.exists(filename):
             os.listdir(os.path.dirname(exit_fn)) # sync NFS
             if not os.path.exists(exit_fn):
                 # (rechecked exit_fn to avoid race condition)
                 fc_run_logger.info( "%r not found, but job is done." % (exit_fn) )
                 fc_run_logger.debug( " job: %r exited." % (job_name) )
-                break
+                return Waiter.FoundOnlyDone
         if task.shutdown_event is not None and task.shutdown_event.is_set():
             fc_run_logger.warning( "shutdown_event received (Keyboard Interrupt maybe?), %r not finished."
                 % (job_name) )
-            if support.job_type == "SGE":
-                fc_run_logger.info( "deleting the job by `qdel` now..." )
-                system("qdel %s" % job_name) # Failure is ok.
-            if support.job_type == "SLURM":
-                fc_run_logger.info( "Deleting the job by 'scancel' now...")
-                system("scancel -n %s" % job_name)
+            cancel_job(job_name, support.job_type)
+            return Waiter.Interrupted
+        return Waiter.StillRunning
+
+def wait_for_file(filename, task, job_name):
+    """(This is the old way.)
+    We could be in the thread or sub-process which spawned a qsub job,
+    so we must check for the shutdown_event.
+    """
+    wait_time = .2
+    while 1:
+        time.sleep(wait_time)
+        wait_time = 10 if wait_time > 10 else wait_time + .2
+        status = check_for_file_or_exit(filename, task, job_name)
+        if status != Waiter.StillRunning:
+            break
+
+def wait_for_file_to_exist(fn):
+    """This can wait forever.
+    """
+    wait_time = .2
+    while 1:
+        time.sleep(wait_time)
+        wait_time = 10 if wait_time > 10 else wait_time + .2
+        if os.path.exists(fn):
+            break
+
+def wait_for_file_with_progress(filename, progress_fn, task, job_name):
+    """Continue as long as we detect progress.
+    Ignore the .exit file. Instead, watch for changes to progress_fn.
```

Comment:
Amendment. This version is just like the old waiter, except that it *also* quits if there is no progress.